text/0800-ts-adoption-plan.md

```diff
@@ -539,36 +539,42 @@ For the classic DI system, we no longer require (or benefit from) string key-bas
+Future designs for services and dependency injection more generally will need to be written to account for the capabilities of TypeScript’s implementation of Stage 3 decorators as they grow and change over time.
+
+For Ember Data, there is considerably more ongoing need for registry-style APIs. `Store.findRecord` isn’t going anywhere any time soon, for example, and it *requires* some kind of registry to make `this.store.findRecord('user', 1)` correctly return a `User` model. Many other APIs within Ember Data similarly require registries for type safety. Thus, Ember Data will need its own dedicated design for handling those, and this RFC leaves those for the Ember Data team to address in a dedicated RFC.
 
 
 ### CLI Integration
 
 We will introduce a new `--typescript` (`-ts`) flag for the `ember new` and `ember addon` commands, allowing users to opt into TypeScript when generating a new project. Using this flag will:
 
-- Install Ember TypeScript integration tooling, today consisting simply of `ember-cli-typescript` and its generators.
 - Set the `isTypeScriptProject` option for `.ember-cli` (introduced in [RFC #0776][rfc-0776]) to `true`, so that blueprints are generated with TypeScript by default in the project.
 - Configure linting:
     - In the`.eslintrc.js` blueprint:
-        - use `@typescript-eslint/parser` instead of the Babel ESLint parser
-        - include `@typescript-eslint` in the `plugins` key
-        - include `plugin:@typescript-eslint/recommended` in the `extends` key
-    - Install the `@typescript-eslint` dependencies instead of of the Babel ESLint dependencies in `package.json`
+        - Use `@typescript-eslint/parser` instead of the Babel ESLint parser.
+        - Include `@typescript-eslint` in the `plugins` key.
+        - Include `plugin:@typescript-eslint/recommended` in the `extends` key.
+    - Install the `@typescript-eslint` dependencies instead of,  or in addition to (as appropriate), the Babel ESLint dependencies in `package.json`.
+- Configure [Glint][glint] for type checking and, for addons, emitting type declarations during build.
+    - Create `tsconfig.json` files and generate their `compilerOptions.paths`.
+    - Configure `ember-cli-babel` to include the TypeScript transform. (This may just be an “out of the box” setting so it always works and does not require configuration!)
```

Comment:
I _think_ the current status quo is that TS transpilation is enabled out-of-the-box with Embroider, but for classic `ember-cli-babel` you still have to opt in via `enableTypeScriptTransform: true` config.

CLI folks, what would you think of changing that default to `true` in the v8 beta?

Reply:
> I think the current status quo is that TS transpilation is enabled out-of-the-box with Embroider

Is this true?
I was setting up TS in an Embroider project, and without the flag, I'm getting errors like:

```ts
$TMPDIR/embroider/9437f6/routes/application.ts/application.ts: Unexpected token, expected "from" (4:12)

  2 | import Route from '@ember/routing/route';
  3 | import { service } from '@ember/service';
> 4 | import type IntlService from 'ember-intl/services/intl';
    |             ^
  5 | import config from 'ember-project-boilerplate/config/environment';
  6 |
  7 | export default class ApplicationRoute extends Route {
```

It could very well be that I'm misconfiguring something though!

Reply:
> CLI folks, what would you think of changing that default to true in the v8 beta?

Seems okay to me, though I don't know if this would have any downsides tbh.

Reply:
@bertdeblock Ahhh, I had tested that `.ts` files were processed, but I didn't actually include any TypeScript-specific syntax in them. As of https://github.com/embroider-build/embroider/pull/1236 it seems like Embroider unconditionally includes `.ts` source files, but still only actually treats it as TypeScript if `'ember-cli-babel': { includeTypeScriptTransform: true }` is set in the app or v1 addon's options.

> I don't know if this would have any downsides for non-TS projects tbh

Outside of some odd edge cases (like if a project has `.ts` and `.js` files with the same name for some reason, or if an addon is erroneously shipping `.ts` files in its `app/` directory), I think this change should be largely invisible for non-TS projects.

The other scenario this might impact would be if someone is manually configuring the TS transform in their Babel config themselves—in that case, they'd need to explicitly opt out with `enableTypeScriptTransform: false` to continue doing that. If this lands as part of the v8 changeover, though, that seems ok to me.

Reply:
If we think there's no downside to enabling it for everyone, then there shouldn't be an option at all. It should just be hard-coded to enabled.

On the other hand if there are legitimate reasons to not want it in an app that doesn't use typescript, and if the default blueprint doesn't use typescript, then I think the default should stay off.

It's true that embroider has some baked-in typescript support, but that is an artifact of our compatibility layers. Where we're heading, your babel config would be completely transparent and it will either have the typescript preset in it or not -- there's no policy question about ember-cli-babel flags.

Reply:
Did a quick test with enabling `enableTypeScriptTransform` in a non-TS project.
TS code in `.js` files is also transpiled away and doesn't error.

Reply:
I think we should have default on, because it makes choosing TS after a JS project easier

Reply:
Having it on by default would be easier indeed. However, it does feel a bit strange that in a non-TS app (where there's no TS related config whatsoever) TS code in a JS file would now also be transpiled away, instead of throwing an error.

Reply:
> On the other hand if there are legitimate reasons to not want it in an app that doesn't use typescript, and if the default blueprint doesn't use typescript, then I think the default should stay off.

As I said above, my concern isn't apps that don't use TS, but rather apps that _are_ using TS but with other nonstandard Babel config that necessitates manual configuration.

That said, now that we've clarified that Embroider only supports the `.ts` extension and not the syntax by default, I'm not as concerned with changing `ember-cli-babel`'s default.